exporter/exporterhelper/internal/queuebatch/persistent_queue.go

```diff
@@ -238,6 +244,140 @@ func (pq *persistentQueue[T]) Offer(ctx context.Context, req T) error {
+	traceID, err := trace.TraceIDFromHex(wrapper.TraceID)
+	if err != nil {
+		return nil, err
+	}
+	spanID, err := trace.SpanIDFromHex(wrapper.SpanID)
+	if err != nil {
+		return nil, err
+	}
+	traceFlags, err := traceFlagsFromHex(wrapper.TraceFlags)
+	if err != nil {
+		return nil, err
+	}
+	traceState, err := trace.ParseTraceState(wrapper.TraceState)
+	if err != nil {
+		return nil, err
+	}
+
+	sc := spanContext(trace.NewSpanContext(trace.SpanContextConfig{
+		TraceID:    traceID,
+		SpanID:     spanID,
+		TraceFlags: *traceFlags,
+		TraceState: traceState,
+		Remote:     wrapper.Remote,
+	}))
+
+	return &sc, nil
+}
+
+// unmarshalRequestWithSpanContext unmarshals a binary envelope, returning the request and a context with the restored SpanContext (if present).
+func unmarshalRequestWithSpanContext[T any](encoding Encoding[T], value []byte) (T, context.Context, error) {
+	var req T
+	restoredContext := context.Background()
+	if len(value) < 8 {
+		return req, restoredContext, errors.New("envelope too short")
+	}
+	reqLen := binary.LittleEndian.Uint32(value[:4])
+	if len(value) < int(4+reqLen+4) {
+		return req, restoredContext, errors.New("envelope too short for request")
+	}
+	reqBytes := value[4 : 4+reqLen]
+	scLen := binary.LittleEndian.Uint32(value[4+reqLen : 8+reqLen])
+	if len(value) < int(8+reqLen+scLen) {
+		return req, restoredContext, errors.New("envelope too short for span context")
+	}
+	scBytes := value[8+reqLen : 8+reqLen+scLen]
+	// Unmarshal request
+	r, err := encoding.Unmarshal(reqBytes)
+	if err != nil {
+		return req, restoredContext, err
+	}
+	req = r
+	// Unmarshal span context if present
+	if scLen > 0 {
+		var wrapper spanContextConfigWrapper
+		if err := json.Unmarshal(scBytes, &wrapper); err == nil {
+			if sc, err := spanContextFromWrapper(wrapper); err == nil && sc != nil {
+				restoredContext = trace.ContextWithSpanContext(restoredContext, trace.SpanContext(*sc))
+			}
+		}
+	}
+	return req, restoredContext, nil
+}
+
+// marshalRequestWithSpanContext marshals the request and the SpanContext from ctx into a binary envelope as bytes.
+func marshalRequestWithSpanContext[T any](ctx context.Context, encoding Encoding[T], req T) ([]byte, error) {
+	reqBuf, err := encoding.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+	sc := trace.SpanContextFromContext(ctx)
+	var scJSON []byte
+	if sc.IsValid() {
+		scJSON, err = json.Marshal(sc)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if len(reqBuf) > int(math.MaxInt32) {
+		return nil, fmt.Errorf("request too large to encode: %d bytes", len(reqBuf))
+	}
+	if len(scJSON) > int(math.MaxInt32) {
+		return nil, fmt.Errorf("span context too large to encode: %d bytes", len(scJSON))
+	}
+	// Compose binary envelope: [4 bytes reqLen][req][4 bytes scLen][scJSON]
+	buf := make([]byte, 0, 8+len(reqBuf)+len(scJSON))
+	//nolint:gosec // G115: integer overflow conversion int -> uint32
+	buf = binary.LittleEndian.AppendUint32(buf, uint32(len(reqBuf)))
+	buf = append(buf, reqBuf...)
```

Comment:
Ok this still requires whole byte array reallocation which is a huge performance hit. I was under impression that we can marshal request into a preallocated array, but apparently pdata doesn't expose that interface. I see there is a another request for that https://github.com/open-telemetry/opentelemetry-collector/issues/12395. Until that capability is introduced, we can go with the two separate keys under alpha feature gate. Once https://github.com/open-telemetry/opentelemetry-collector/issues/12395 is resolved, we should be able to get back to this approach with performance improvement.

Reply:
But before jumping on https://github.com/open-telemetry/opentelemetry-collector/issues/12395, we should probably export and use https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/internal/data/protogen/trace/v1/trace.pb.go#L951 this PR to see the benchmark numbers that we would get eventually. If you could do that whenever you have time, that would be perfect.